rts/idris_rts.c

```diff
@@ -411,102 +435,186 @@ void idris_memmove(void* dest, void* src, i_int dest_offset, i_int src_offset, i
         exit(EXIT_FAILURE);
     }
 
-    SETTY(cl, STRING);
-    return cl;
+    // avoid invalid UTF-8: all ASCII digits are acceptable as UTF-8 characters.
+    int j;
+    for (j = 0; j < 21; j++) {
+        if (temp[j] != '\0' && !(isdigit(temp[j]))) {
+            fprintf(stderr, "Fatal error: invalid output from sprintf on bit type %d!", ty);
+            exit(EXIT_FAILURE);
+        }
+    }
+
+    return MKSTR(vm, (const utf8_byte *)temp);
 }
 
 VAL idris_castStrInt(VM* vm, VAL i) {
     char *end;
-    i_int v = strtol(GETSTR(i), &end, 10);
-    if (*end == '\0' || *end == '\n' || *end == '\r')
-        return MKINT(v);
+    SizedString str = GETSTR(i);
+
+    // First, we need to alloc a C string that's long enough (to get null-termination)
+    char *temp = (char *) malloc(str.size + 1);
+    if (temp == NULL) {
+        fprintf(stderr, "Couldn't allocate temp buffer of size %lu for int->string conversion", str.size);
+        exit(EXIT_FAILURE);
+    }
+    // Then grab the UTF8 bytes out of the Idris string. We don't have
+    // to worry about nulls in the middle because we'll later check to
+    // see if the whole string was used.
+    memcpy(temp, str.str_data, str.size);
+    temp[str.size] = '\0';
+
+    i_int v = strtol(temp, &end, 10);
+    Closure* answer;
+    // If we used the whole string, return the int, otherwise return 0. This is
+    // the meaning of the primitive in the evaluator.
+    if (end == temp + str.size)
+        answer = MKINT(v);
     else
-        return MKINT(0);
+        answer = MKINT(0);
+    free(temp);
+    return answer;
 }
 
 VAL idris_castFloatStr(VM* vm, VAL i) {
-    Closure* cl = allocate(sizeof(Closure) + sizeof(char)*32, 0);
-    SETTY(cl, STRING);
-    cl -> info.str = (char*)cl + sizeof(Closure);
-    snprintf(cl -> info.str, 32, "%.16g", GETFLOAT(i));
-    return cl;
+    char temp[32] = { 0 };
+    // snprintf will put a null at the end, so 31 chars are available for the float
+    snprintf(temp, 32, "%.16g", GETFLOAT(i));
+
+    // avoid invalid UTF-8: all ASCII digits, plus period, plus, and
+    // minus, are acceptable as UTF-8 characters.
+    int j;
+    for (j = 0; j < 21; j++) {
+        if (temp[j] != '\0' &&
+            temp[j] != '.' &&
+            temp[j] != '-' &&
+            temp[j] != '+' &&
+            !(isdigit(temp[j]))
+           ) {
+            fprintf(stderr, "Fatal error: invalid output from sprintf was %s!", temp);
+            exit(EXIT_FAILURE);
+        }
+    }
+
+
+    return MKSTR(vm, (utf8_byte *)temp);
 }
 
 VAL idris_castStrFloat(VM* vm, VAL i) {
-    return MKFLOAT(vm, strtod(GETSTR(i), NULL));
+    SizedString str = GETSTR(i);
+
+    // First, we need to alloc a C string that's long enough (to get null-termination)
+    char *temp = (char *) malloc(str.size + 1);
+    if (temp == NULL) {
+        fprintf(stderr, "Couldn't allocate temp buffer of size %lu for int->string conversion", str.size);
```

Comment:
I don'rt see why you'd need a custom message if malloc fails. It's a system thing, the only thing that could be of interest would be the size of the allocation. It's an extreme corner case anyhow, and it's very unlikely the system will live long enough to print anything if small allocs fail.

Many programs don't bother to handle alloc failure at all (and OpenBSD's malloc just aborts the program). It's an extreme corner case, the least we can do is to handle it in one place and not sprinkle error handling for it all over.